src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/AddUsingsCodeActionProviderHelper.cs

```diff
@@ -0,0 +1,54 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.LanguageServer.CodeActions.Models;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using OmniSharp.Extensions.LanguageServer.Protocol;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.CodeActions
+{
+    internal static class AddUsingsCodeActionProviderHelper
+    {
+        internal static string GetNamespaceFromFQN(string fullyQualifiedName)
```

Comment:
> Are you sure translating a FQN into an Add using is always correct 

No.

> understanding when C# provides the using light bulb?

So this was the alternative. It seemed the solution here would likely add a fair bit of complexity, given we have to make changes to how we handle the code behind as @ryanbrandenburg detailed [here](https://github.com/dotnet/aspnetcore/issues/18173). The TLDR is we have `#line hidden`s around the `using` block at the start of the generated document, as well as some near the code block area. This results in Roslyn not providing any `using` code action for us. 

I'm not sure how much of an issue this would present in practice / whether it's worth prioritizing over other changes / code actions. Thoughts?

> If the assumption is wrong we could potentially be offering the add using light bulb when the using statement already exists and invoking the light bulb would result in the Razor document being invalid.

If we re-add a using statement that already exists, it shouldn't matter(?) as the compiler just ignores duplicate `using`. Also, as this is just doing a `@using` at the top of the razor, unexpected behavior can quickly be rectified with a undo (no corruption of background document or anything).

Reply:
> I'm not sure how much of an issue this would present in practice / whether it's worth prioritizing over other changes / code actions. Thoughts?

To be honest I totally forgot about the Roslyn implication of them not actually providing the using light bulb due to our hidden directives. Because of that what you have here is an awesome alternative. I think eventually we'll probably want to work with Roslyn to enable those light bulbs in Razor files despite the line directives but there's a lot of value in seeing how far the solution you have here can get users. Definitely something we can re-evaluate in the future based on feedback.



> If we re-add a using statement that already exists, it shouldn't matter(?) as the compiler just ignores duplicate `using`. Also, as this is just doing a `@using` at the top of the razor, unexpected behavior can quickly be rectified with a undo (no corruption of background document or anything).

Technically it shouldn't matter; however, if adding that using statement then makes other pieces of the document "ambiguous" adding the using statement could cause errors. That being said, it's not a common case scenario so it's definitely worth ignoring for the time being.